stdlib/asm/collections/mmr.masm

```diff
@@ -104,3 +106,76 @@ export.get
      # stack: [leaf, ...]
   end
 end
+
+#! Load the MMR peak data based on its hash.
+#!
+#! Input: [HASH, mmr_ptr, ...]
+#! Output: [...]
+#!
+#! Where:
+#!  - HASH: is the MMR peak hash, the hash is expected to be padded to an even
+#!    length and to have a minimum size of 16 elements
+#!  - mmt_ptr: the memory location where the MMR data will be written to,
+#!    starting with the MMR forest (its total leaves count) followed by its peaks
+#!
+#! Cycles: 159 + 9 extra_peak_pair cycles
+#!    where `extra_peak` is the number of peak pairs in addition to the first
+#!    16, i.e. `round_up((num_of_peaks - 16) / 2)`
+export.unhash
+  # load the peaks to the advice_stack (0 cycles)
+  adv.keyval
+  # operand_stack => [HASH, mmr_ptr, ...]
+  # advice_stack => [peaks*, ...]
+
+  # load the number of leaves in the MMR to the advice_stack, the number of
+  # leaves is defined by the hash of the MMR peaks + 1 (3 cycles)
+  add.1 adv.keyval sub.1
+  # operand_stack => [HASH, mmr_ptr, ...]
+  # advice_stack => [forest, peaks*, ...]
+
+  # load the size from the advice stack (1 cycles)
+  adv_push.1
+  # operand_stack => [forest, HASH, mmr_ptr, ...]
+  # advice_stack => [peaks*, ...]
+
+  # save the forest to memory (4 cycles)
+  dup dup.6 mem_store
+  # => [forest, HASH, mmr_ptr, ...]
+
+  # count number of peaks (69 cycles)
+  u32split u32unchecked_popcnt swap u32unchecked_popcnt add
```

Comment:
I usually use `u32` but on previous PRs it is better to lift such requirements, this was done to handle a Felt instead.

Reply:
The prover could use a number of peaks that is different that the number of leaves, and I'm not sure what problems could raise from that. Somehow we have to validate the two match, I guess we could encode it as:

```
sizes = number_of_leaves1 || number_of_leaves2 || .. || number_of_leavesN
peaks = peak1 || peak2 ... || peakN
number_of_peaks || sizes || peaks
```

And then the MMR `number_of_leaves` would be `sum(sizes)`. But I haven't checked to see if that is more efficient.

Reply:
good catch. I'd like to understand the implications of a malicious prover in this scenario. In any case I think it's better to leave it as it is.